File: bindings/deno/src/safe_cookie.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeCookie - HTTP cookie operations that prevent injection attacks.
 *
 * Thin FFI wrapper: all computation delegates to libproven (Idris 2 + Zig).
 *
 * @module
 */

import { getLib, ProvenStatus, readAndFreeString, statusToError } from './ffi.ts';
import { err, ok, type Result } from './result.ts';

/**
 * Cookie prefix type constants.
 */
export const CookiePrefix = {
  NONE: 0,
  SECURE: 1,
  HOST: 2,
} as const;

/**
 * Safe cookie operations backed by formally verified Idris 2 code.
 */
export class SafeCookie {
  /**
   * Check for cookie injection characters (semicolon, CR, LF).
   *
   * @param value - The value to check.
   * @returns Result containing a boolean or an error string.
   */
  static hasInjection(value: string): Result<boolean> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(value);
    const result = symbols.proven_cookie_has_injection(bytes, bytes.length);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(result[1]);
  }

  /**
   * Validate a cookie name.
   *
   * @param name - The cookie name.
   * @returns Result containing a boolean or an error string.
   */
  static validateName(name: string): Result<boolean> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(name);
    const result = symbols.proven_cookie_validate_name(bytes, bytes.length);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(result[1]);
  }

  /**
   * Validate a cookie value.
   *
   * @param value - The cookie value.
   * @returns Result containing a boolean or an error string.
   */
  static validateValue(value: string): Result<boolean> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(value);
    const result = symbols.proven_cookie_validate_value(bytes, bytes.length);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(result[1]);
  }

  /**
   * Get cookie prefix type (0=none, 1=__Secure-, 2=__Host-).
   *
   * @param name - The cookie name.
   * @returns Result containing the prefix type or an error string.
   */
  static getPrefix(name: string): Result<number> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(name);
    const result = symbols.proven_cookie_get_prefix(bytes, bytes.length);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(Number(result[1]));
  }

  /**
   * Build a delete cookie header value (expired Set-Cookie).
   *
   * @param name - The cookie name to delete.
   * @returns Result containing the Set-Cookie header or an error string.
   */
  static buildDelete(name: string): Result<string> {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(name);
    const result = symbols.proven_cookie_build_delete(bytes, bytes.length);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    const header = readAndFreeString(result[1], Number(result[2]));
    if (header === null) return err('Null string returned');
    return ok(header);
  }
}

File: bindings/swift/Sources/Proven/SafeCookie.swift
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// Safe HTTP cookie operations delegated to libproven FFI.
///
/// All validation and rendering is performed by the formally verified
/// Idris 2 core via the Zig FFI bridge.

import CProven

public enum SafeCookie {
    /// Check if a cookie value contains injection characters.
    public static func hasInjection(_ value: String) -> Result<Bool, ProvenError> {
        withStringBytes(value) { ptr, len in
            let result = proven_cookie_has_injection(ptr, len)
            if let error = ProvenError.fromStatus(result.status) {
                return .failure(error)
            }
            return .success(result.value)
        }
    }

    /// Validate a cookie name per RFC 6265.
    public static func validateName(_ name: String) -> Result<Bool, ProvenError> {
        withStringBytes(name) { ptr, len in
            let result = proven_cookie_validate_name(ptr, len)
            if let error = ProvenError.fromStatus(result.status) {
                return .failure(error)
            }
            return .success(result.value)
        }
    }

    /// Validate a cookie value per RFC 6265.
    public static func validateValue(_ value: String) -> Result<Bool, ProvenError> {
        withStringBytes(value) { ptr, len in
            let result = proven_cookie_validate_value(ptr, len)
            if let error = ProvenError.fromStatus(result.status) {
                return .failure(error)
            }
            return .success(result.value)
        }
    }

    /// Get the security prefix type of a cookie name.
    /// Returns an integer code identifying the prefix (e.g. __Secure-, __Host-).
    public static func getPrefix(_ name: String) -> Result<Int64, ProvenError> {
        withStringBytes(name) { ptr, len in
            let result = proven_cookie_get_prefix(ptr, len)
            if let error = ProvenError.fromStatus(result.status) {
                return .failure(error)
            }
            return .success(result.value)
        }
    }

    /// Build a Set-Cookie header string from validated components.
    public static func buildSetCookie(
        name: String,
        value: String,
        maxAge: Int64 = 0,
        secure: Bool = true,
        httpOnly: Bool = true,
        sameSiteStrict: Bool = true
    ) -> Result<String, ProvenError> {
        let nameUtf8 = Array(name.utf8)
        let valueUtf8 = Array(value.utf8)
        return nameUtf8.withUnsafeBufferPointer { nameBuf in
            valueUtf8.withUnsafeBufferPointer { valueBuf in
                let result = proven_cookie_build_set_cookie(
                    nameBuf.baseAddress, nameBuf.count,
                    valueBuf.baseAddress, valueBuf.count,
                    maxAge, secure, httpOnly, sameSiteStrict
                )
                return consumeStringResult(result)
            }
        }
    }

    /// Build a Set-Cookie header that deletes a cookie.
    public static func buildDelete(name: String) -> Result<String, ProvenError> {
        withStringBytes(name) { ptr, len in
            consumeStringResult(proven_cookie_build_delete(ptr, len))
        }
    }
}
